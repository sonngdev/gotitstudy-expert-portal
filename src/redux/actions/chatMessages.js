export const CHAT_MESSAGES = {
  ADD: 'CHAT_MESSAGES/ADD',
}

export function chatMessagesAdd(message) {
  return { type: CHAT_MESSAGES.ADD, message };
}
