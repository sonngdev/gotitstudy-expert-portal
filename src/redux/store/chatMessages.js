export const chatMessagesInitState = [
  {
    type: 'reviser',
    avatar: 'asker',
    text: 'Please help me',
  },
  {
    type: 'system',
    avatar: 'gotit',
    text: 'GotIt! monitors chat for the safety of our community. Never share any personal or private information. Please use the flag button to notify the GotIt! Team of inapporpriate behavior.',
  },
];
