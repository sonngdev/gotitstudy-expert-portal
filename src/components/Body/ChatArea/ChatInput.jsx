import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { chatMessagesAdd } from '../../../redux/actions/chatMessages';
import { Button, Form } from '@gotitinc/design-system';

export function ChatInput() {
  const [ input, setInput ] = useState('');
  const dispatch = useDispatch();
  const bindInput = (e) => setInput(e.target.value);
  const sendMessage = (e) => {
    e.preventDefault();
    const msg = { type: 'sender', avatar: 'expert', text: input };
    const action = chatMessagesAdd(msg);
    dispatch(action);
    setInput('');
  }

  return (
    <Form onSubmit={sendMessage}>
      <Form.InputGroup>
        <Form.Input placeholder="Message..."
          name="message"
          autocomplete="off"
          value={input}
          onChange={bindInput}
        />
        <Form.InputGroup.Append className="u-flex">
          <Button>
            <Button.Label>Send</Button.Label>
          </Button>
        </Form.InputGroup.Append>
      </Form.InputGroup>
    </Form>
  );
}
