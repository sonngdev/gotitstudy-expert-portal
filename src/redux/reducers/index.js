import { combineReducers } from 'redux';
import { chatMessagesReducer } from './chatMessages';

export const rootReducer = combineReducers({
  chatMessages: chatMessagesReducer,
});
