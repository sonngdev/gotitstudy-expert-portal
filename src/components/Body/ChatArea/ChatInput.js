import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { Button, Form } from '@gotitinc/design-system';
import { chatMessagesAdd } from '../../../redux/actions';

export function ChatInput() {
  const [input, setInput] = useState('');
  const dispatch = useDispatch();
  const bindInput = (e) => setInput(e.target.value);
  const sendMessage = (e) => {
    e.preventDefault();

    const trimmedInput = input.trim();
    if (!trimmedInput) return;

    const msg = { type: 'sender', avatar: 'expert', text: trimmedInput };
    const action = chatMessagesAdd(msg);
    dispatch(action);
    setInput('');
  };

  return (
    <Form onSubmit={sendMessage} className="u-paddingRightSmall">
      <Form.InputGroup>
        <Form.Input
          placeholder="Message..."
          name="message"
          autoComplete="off"
          autoFocus
          value={input}
          onChange={bindInput}
        />
        <Form.InputGroup.Append className="u-flex">
          <Button>
            <Button.Label>Send</Button.Label>
          </Button>
        </Form.InputGroup.Append>
      </Form.InputGroup>
    </Form>
  );
}
