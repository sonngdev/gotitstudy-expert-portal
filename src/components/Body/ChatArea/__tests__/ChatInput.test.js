import React from 'react';
import { Form } from '@gotitinc/design-system';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { store } from '../../../../redux/store';
import { chatMessagesAdd } from '../../../../redux/actions';
import { ChatInput } from '../ChatInput';

describe('<ChatInput />', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(
      <Provider store={store}>
        <ChatInput />
      </Provider>,
    );
  });

  it('dispatches add message action when user submit message', () => {
    const spy = jest.spyOn(store, 'dispatch');
    spy.mockImplementation(jest.fn());

    const message = 'Hello, how can I help?';
    const input = wrapper.find(Form.Input);
    const form = wrapper.find(Form);
    const payload = chatMessagesAdd({
      type: 'sender',
      avatar: 'expert',
      text: message,
    });

    input.simulate('change', { target: { value: message } });
    form.simulate('submit');

    expect(spy).toHaveBeenCalledWith(payload);

    spy.mockRestore();
  });
});
