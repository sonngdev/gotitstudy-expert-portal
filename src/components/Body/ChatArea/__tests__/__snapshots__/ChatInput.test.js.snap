// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatInput /> renders 1`] = `
"<ForwardRef onSubmit={[Function: sendMessage]} className=\\"u-paddingRightSmall\\">
  <Form.InputGroup>
    <Form.Input placeholder=\\"Message...\\" name=\\"message\\" autoComplete=\\"off\\" autoFocus={true} value=\\"\\" onChange={[Function: bindInput]} />
    <FormInputGroupAppend uFlex className=\\"u-flex\\">
      <Button nonUppercase={false} variant=\\"primary\\">
        <ButtonLabel uInlineBlock>
          Send
        </ButtonLabel uInlineBlock>
      </Button>
    </FormInputGroupAppend uFlex>
  </Form.InputGroup>
</ForwardRef>"
`;
